Loosen some lifetime dependence diagnostics

meg-gupta opened the pull request:
This change removes diagnostic for lifetime dependence on ~Escapable & _BitwiseCopyable type.

Also includes changes to _BitwiseCopyable to be ~Escapable when the NonEscapableTypes feature is enabled.

And changes that add build-script option and lit config for building the compiler and testing with NonEscapableTypes feature.

rdar://123870480

meg-gupta committed:
Tighten condition under which we bailout from lifetime dependence inference of implicit functions

meg-gupta committed:
Dont diagnose lifetime dependence on BitwiseCopyable  & ~Escapable type in Sema

rdar://123870480

meg-gupta commented:
@swift-ci smoke test

meg-gupta commented:
@swift-ci smoke test

atrick commented on lib/Sema/LifetimeDependence.cpp:
```diff
@@ -159,12 +159,19 @@ LifetimeDependenceInfo::fromTypeRepr(AbstractFunctionDecl *afd, Type resultType,
-        diags.diagnose(loc, diag::lifetime_dependence_on_bitwise_copyable);
-        return true;
+    // Diagnose when we have lifetime dependence on a type that is
+    // BitwiseCopyable & Escapable ~Escapable types are non-trivial in SIL and
```
> BitwiseCopyable & Escapable ~Escapable types

Missing punctuation

meg-gupta replied:
Done.

atrick commented on test/Sema/explicit_lifetime_dependence_specifiers.swift:
```diff
@@ -6,6 +6,14 @@ struct Container {
   let ptr: UnsafeRawBufferPointer
 }
 
+struct AnotherBufferView : ~Escapable {
```
I AnotherBufferView inferred to be BitwiseCopyable? Do you want to specify BitwiseCopyable here to be explicit?

atrick reviewed (approved):
LGTM. 2 comments...

meg-gupta committed:
Remove diagnostic which didnt allow explicitly specifying BitwiseCopyable on a ~Escapable type

meg-gupta committed:
Allow ~Escapable types to be _BitwiseCopyable

meg-gupta committed:
Add build-script option and lit config for nonescapable_types

With this build-script has a flag --enable-experimental-nonescpable-types=1
to enable this feature in stdlib.

Also we can now add // REQUIRES: nonescapable_types to tests which run only when
the compiler is built with this feature turned on.

meg-gupta commented:
@swift-ci test

meg-gupta commented:
@swift-ci test

nate-chandler reviewed (approved):
Thank you!

meg-gupta commented:
@swift-ci smoke test

meg-gupta commented:
@swift-ci smoke test
